src/sdk/pynni/nni/ppo_tuner/ppo_tuner.py

```diff
@@ -0,0 +1,574 @@
+                one_mask[idx] = 0
+            mask.append(one_mask)
+        two_masks.append(mask)
+
+        return np.asarray(two_masks, dtype=np.float32)
+
+    def update_search_space(self, search_space):
+        """
+        get search space, currently the space only includes that for NAS
+
+        Parameters:
+        ----------
+        search_space: json object                 search space for NAS
+        
+        Returns:
+        -------
+        no return
+        """
+        logger.info('=== update search space %s', search_space)
+        assert self.search_space is None
+        self.search_space = search_space
+
+        assert self.model_config.observation_space is None
+        assert self.model_config.action_space is None
+
+        self.actions_spaces, self.actions_to_config, self.full_act_space, obs_space, nsteps = self._process_nas_space(search_space)
+        
+        self.model_config.observation_space = spaces.Discrete(obs_space)
+        self.model_config.action_space = spaces.Discrete(obs_space)
+        self.model_config.nsteps = nsteps
+
+        # generate mask in numpy
+        mask = self._generate_action_mask()
+
+        assert self.model is None
+        self.model = PPOModel(self.model_config, mask)
+
+    def _trial_dedup(self):
+        pass
+
+    def _actions_to_config(self, actions):
+        """
+        given actions, to generate the corresponding trial configuration
+        """
+        import copy
+        chosen_arch = copy.deepcopy(self.chosen_arch_template)
+        self.actions_spaces, self.actions_to_config, self.full_act_space
+        for cnt, act in enumerate(actions):
+            act_name = self.full_act_space[act]
+            idx = self.actions_spaces[cnt].index(act_name)
+            (block_name, layer_name, key) = self.actions_to_config[cnt]
+            if key == 'chosen_inputs':
+                if act_name == 'None':
+                    chosen_arch[block_name][layer_name][key] = []
+                else:
+                    chosen_arch[block_name][layer_name][key] = [act_name]
+            elif key == 'chosen_layer':
+                chosen_arch[block_name][layer_name][key] = act_name
+            else:
+                raise ValueError("unrecognized key: %s".format(key))
+        return chosen_arch
+
+    def generate_multiple_parameters(self, parameter_id_list, **kwargs):
+        """
+        Returns multiple sets of trial (hyper-)parameters, as iterable of serializable objects.
+        """
+        result = []
+        for parameter_id in parameter_id_list:
+            had_exception = False
+            try:
+                logger.debug("generating param for {}".format(parameter_id))
+                res = self.generate_parameters(parameter_id, **kwargs)
+            except nni.NoMoreTrialError:
+                had_exception = True
+            if not had_exception:
+                result.append(res)
+        return result
```

Comment:
just a question:  is there any case there are new parameters generated after NoMoreTrialError is raised?  
It seems  self.trails_info.get_next() only return None when iter > size

Reply:
NoMoreTrialError is not raised (or it is catched, not exposed to msg_dispatcher), because for PPO tuner it can always have new parameters. The situation is that this tuner has to wait all issued trial configs to complete, then generate new configs. That is why I use self.credit to record.